### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by LukeMathWalker** [rust-ndarray/ndarray-stats#1]
In terms of functionality, the mid-term end goal is to achieve feature parity with the statistics routine in `numpy` ([here](https://docs.scipy.org/doc/numpy/reference/routines.statistics.html)) and `Julia StatsBase` ([here](https://juliastats.github.io/StatsBase.jl/latest/)).

For the next version:
- Order statistics:
   - [ ] `partialord` version for `quantiles` methods;
- Histograms:
   - [ ] [`merge`](http://juliastats.github.io/StatsBase.jl/stable/empirical.html#Base.merge) method;

------

For version **0.2.0**:
- Order statistics:
   - [x] optimized computations of multiple quantiles if requested all at once (#26)  ;
   - [x] [argmin](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.argmin.html) / [argmax](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.argmax.html#numpy.argmax) (#30);
- Summary statistics:
   - [x] [harmonic mean](https://en.wikipedia.org/wiki/Harmonic_mean) (#20);
   - [x] [geometric mean](https://en.wikipedia.org/wiki/Geometric_mean) (#20);
   - [x] [higher order central moments](https://en.wikipedia.org/wiki/Central_moment) (#23);
   - [x] [standardized moments](https://en.wikipedia.org/wiki/Standardized_moment) (they include [kurtosis](https://en.wikipedia.org/wiki/Kurtosis) and [skewness](https://en.wikipedia.org/wiki/Skewness)) (#23);
- Histograms:
   - [x] Fix error handling (Issue: https://github.com/jturner314/ndarray-stats/issues/16 - PR: #25 )
- Entropy:
   - [x] Feature parity with [StatsBase.jl](http://juliastats.github.io/StatsBase.jl/stable/scalarstats.html#Entropy-and-Related-Functions-1) (#24)

For version  **0.1.0**:
- [x] `max` / `nanmax` (@jturner314)
- [x] `min` / `nanmin` (@jturner314)
- [x] `quantile` / `nanquantile` (it includes `percentile` / `nanpercentile` as a special case) (@LukeMathWalker & @jturner314)
- [x] `correlation`-methods:
   - [x] `cov` (@LukeMathWalker) - ~One last fix to be made (#3)~ [On hold for now]
   - [x] `corrcoef` (@LukeMathWalker - #5)
- [x] `histogram`-methods (@LukeMathWalker - #9)

**Comment by LukeMathWalker** [comment]
Once #9 gets merged I think we are in a good position to officially release version 0.1.0 on crates.io - what do you think? @jturner314

**Comment by jturner314** [comment]
I agree.

By the way, I recently came across Julia's [`StatsBase.jl` library](https://juliastats.github.io/StatsBase.jl/latest/). It's a good source of ideas in addition to NumPy/SciPy.

**Comment by LukeMathWalker** [comment]
Added a bunch of tests to #9 and merged 🎉 It feels like ages since I started to work on it :sweat_smile: Your contribution was extremely helpful to get it in the shape it is right now, thanks a lot @jturner314!

What do we need to do in order to release on crates.io?
I am going to open a small PR to add crate-level documentation - a couple of lines, nothing major.

**Comment by jturner314** [comment]
Yay! :tada: That was a big job; great work.

> What do we need to do in order to release on crates.io?

Ideally, we'd eliminate the `[patch.crates-io]` section from the `Cargo.toml` before we can release on crates.io. (This might even be required, I'm not sure.) #11 removed the patch for `noisy_float`, but a new version of `ndarray` will need to be released for us to remove its patch. It would be nice to merge a couple more `ndarray` PRs before release; I'll take a look.

It would also be good to merge #12 and #13 before releasing.

**Comment by LukeMathWalker** [comment]
Merged #12 and #13 - looking around it seems we can publish with `[patch.crates-io]` section in `Cargo.toml`, but I agree it is much nicer to point to `ndarray` 0.12.1 as a dependency instead of a revision on `master`.

Let's wait for that release and then we are good to go.

**Comment by jturner314** [comment]
`ndarray-stats` 0.1.0 is now on crates.io. :tada: Thanks for all your hard work @LukeMathWalker!

**Comment by LukeMathWalker** [comment]
💯 💯

**Comment by LukeMathWalker** [comment]
I have drafted a tentative roadmap with the features I'd like to add in the next release - please edit it with your comments and suggestions @jturner314

**Comment by phungleson** [comment]
Hey mate, `argmin / argmax` looks like simple enough to look into, do you have any suggestions of where to start?

**Comment by jturner314** [comment]
Thanks for your interest! You'll want to add `argmin` and `argmax` methods to the [`QuantileExt` trait](https://github.com/jturner314/ndarray-stats/blob/master/src/quantile.rs#L180) and [implement them](https://github.com/jturner314/ndarray-stats/blob/master/src/quantile.rs#L276). Please include documentation for the methods and some tests (in [`tests/quantile.rs`](https://github.com/jturner314/ndarray-stats/blob/master/tests/quantile.rs)).

I'd suggest starting with the existing implementation for [`min`](https://github.com/jturner314/ndarray-stats/blob/master/src/quantile.rs#L281-L290) as a basis, but using [`.indexed_iter().fold()`](https://docs.rs/ndarray/0.12.1/ndarray/struct.ArrayBase.html#method.indexed_iter) or `.indexed_iter().try_fold()` instead of [`.fold()`](https://docs.rs/ndarray/0.12.1/ndarray/struct.ArrayBase.html#method.fold).

It would also be good to add `argmin_skipnan` and `argmax_skipnan` methods (analogous to `min_skipnan` and `max_skipnan`, but that's not necessary for the first PR.

Please feel free to ask if you have any questions.

**Comment by phungleson** [comment]
Hey mates, I have added `argmin_skipnan` and `argmax_skipnan`, wonder why you use `PartialOrd` for `min`, but `Ord` for `min_skipnan`?

And what does this mean by this? `partialord version for quantiles`

**Comment by LukeMathWalker** [comment]
It's because we require the data type to be `MaybeNan`: it basically means that, apart from a subset of elements (e.g. `NaN` for floats), we are dealing with a data type that is totally ordered (all pairs of elements can be compared, `Ord`).

This reduces the failure scope: 
- `min` can return `None` is a comparison fails (as it can happen, with `PartialOrd`) or if there is no element in the array. 
- `min_skipnan` returns `None` if and only if the array has no not-NaN element (because no comparison will be undefined).

Re: quantiles - the current implementation requires `A` to implement `Ord`. We'd like to relax it to allow `A` to be `PartialOrd` instead of `Ord`.

**Comment by phungleson** [comment]
Thanks @LukeMathWalker for the last point, if we change `A: Ord` to `A: PartialOrd` and refactor the code + test to allow that change, it would complete the task right?

**Comment by LukeMathWalker** [comment]
Exactly! @phungleson 
I'd suggest you though to wait until #26 is merged before tackling this task, otherwise you are in for some nasty merge conflicts :stuck_out_tongue:

**Comment by phungleson** [comment]
Cool thanks @LukeMathWalker so seems like everything is more or less complete? Let me know if there are any doable features, cheers.

BTW `merge method;` seems to be straight forward but do you have any thoughts yet about the implementation?

**Comment by phungleson** [comment]
For `merge` I read quickly, so basically just adding the weights?

```
for h in others
  target.weights .+= h.weights
end
```

**Comment by LukeMathWalker** [comment]
Yes @phungleson, it basically boils down to summing together the `weight` matrices (plus or minus checking that their dimension/bins are compatible, I haven't looked into it). If you want to give it try, please go ahead!

**Comment by LukeMathWalker** [comment]
I'd like to close existing work streams and cut a release - what does your bandwidth look like @jturner314 to review open PRs?

**Comment by jturner314** [comment]
I've been meaning to look over the open PRs but haven't had a chance. I'll reserve time on Sunday to review them.

**Comment by LukeMathWalker** [comment]
It seems I managed to publish `0.2.0` without making a mess :muscle: 
Thanks @jturner314 @phungleson and @munckymagik for all the work done on this release :heart: 

I'd say we have done a major leap forward in terms of features - there are things that can be polished, the API design can be further improved and we can optimize the existing code, but `ndarray-stats` is definitely a viable solution right now :rocket: 

I'll clean up the parent post to move items that we didn't manage to include in this release to the roadmap for the next one. I am not sure what we should be covering next in terms of major new functionality :thinking:

**Comment by munckymagik** [comment]
Well done all 👏

**Comment by jturner314** [comment]
Great job on 0.2.0 everyone!

> I am not sure what we should be covering next in terms of major new functionality

A couple of ideas from StatsBase.jl:

* [Deviation functions](https://juliastats.github.io/StatsBase.jl/latest/deviation.html)
* [Weighted calculations](https://juliastats.github.io/StatsBase.jl/latest/weights.html) (mean/std/etc.)

We could also add statistical models (e.g. linear regression), but that might be best put in a separate crate.

**Comment by phungleson** [comment]
Well done! cheers!

**Comment by munckymagik** [comment]
> A couple of ideas from StatsBase.jl:
>
>  *  Deviation functions
>  *  Weighted calculations (mean/std/etc.)

Unless any of you have made a start on these, I'd be interested in having a go at either, or contributing. I'll try to spend some time in the next couple of days looking at what is involved with the Deviation functions.

❓ Does anyone have any implementation suggestions other than just trying to port from StatsBase.jl?

If anyone wants to collaborate on the code then let me know.

**Comment by munckymagik** [comment]
Ok I made a start: https://github.com/jturner314/ndarray-stats/pull/41

Any advice for choosing traits bounds for the `A` element types? Is it ok to use `Copy` or do we need to support any types that would be `Clone`?

**Comment by LukeMathWalker** [comment]
I'd say to use `clone`

**Comment by munckymagik** [comment]
@LukeMathWalker thanks. What led you to that decision? Is there a particular data type you've seen used in `ndarray`s that would need this? If so I'm thinking I might use it in the test fixtures to make sure all methods have the same bounds.

**Comment by LukeMathWalker** [comment]
I see it as a tradeoff between convenience and generality - I am not personally aware of any "popular" numerical type that is not `Copy`, but the cost of weakening it to `Clone` is so low that I see it as safe future-proofing @munckymagik

**Comment by nilgoyette** [comment]
I wanted to code a simple `weighted_mean` for myself then contribute it

    pub fn weighted_mean<A, S>(data: &ArrayBase<S, Ix1>, weights: &[A]) -> A
    where
        S: Data<Elem = A>,
        A: Float,
    {
        data.iter().zip(weights).fold(A::zero(), |acc, (&d, &w)| acc + d * w)
    }

but I realize that it's too simple. This code is only useful for 1D arrays, or flattened matrices/images, etc. I can change the `Ix1` for a `D: Dimension`, so that we don't need to flatten anything. It's still a one-liner though and it doesn't offer any "axis" feature, like [Numpy](https://docs.scipy.org/doc/numpy/reference/generated/numpy.average.html). I think we need 2 functions here, because they won't return the same type.

1. n-d data with n-d weight, returns a number
2. n-d data with 1d weight + axis, returns (n-1)-d array.

What do you guys had in mind?

**Comment by LukeMathWalker** [comment]
> I think we need 2 functions here, because they won't return the same type.
> 
> 1. n-d data with n-d weight, returns a number
> 2. axis mode: n-d data with n-d weight, returns (n-1)-d array.
> 
> What do you guys had in mind?

I think it makes perfect sense to have two functions.
It is also consistent with the rest of the API: we have `mean` and `mean_axis`, `var` and `var_axis`, etc. :+1:

**Comment by aeroaks** [comment]
I was thinking of picking up the histogram [`merge`](http://juliastats.github.io/StatsBase.jl/stable/empirical.html#Base.merge) method.
I am relatively new to rust and `ndarray`. With this exercise, I want to pick up ndarray, rust and also start contributing to `ndarray-*` libraries.
What do you guys think? Or are there more high level `good first issue` in other `ndarray-*` libraries?

**Comment by munckymagik** [comment]
@aeroaks I'd say go for it 💯 

You could raise a draft PR if you get something working and want some early feedback.

**Comment by RolfStierle** [comment]
I would like to implement something like [scipy.stats.binned_statistic_dd](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.binned_statistic_dd.html#scipy.stats.binned_statistic_dd) based on [ndarray_stats::histogram::Histogram](https://docs.rs/ndarray-stats/0.3.0/ndarray_stats/histogram/struct.Histogram.html), allowing to caluclate running means, variances, sums, max/min value in each bin. 
Would that be of interest?

**Comment by LukeMathWalker** [comment]
It does! @RolfStierle

**Comment by lebensterben** [comment]
According to https://docs.scipy.org/doc/numpy/reference/generated/numpy.histogram_bin_edges.html#numpy.histogram_bin_edges

Some of bins building strategies are not implemented by `rust-ndarray` now:

- `doane`
- `scott`
- `stone`

**Comment by humphreylee** [comment]
Thanks very much for sharing the good work. Would it be possible to add univariate, bivariate and multivariate kernel density estimation functions? Thanks.
